src/provider/sl/sl.go

```diff
@@ -48,6 +49,139 @@ func New(conf *SLConfig) (*SLImages, error) {
 	}, nil
 }
 
+// Transaction returns an ongoing transaction for the image given by the id.
+//
+// It returns non-nil error when querying the service failed.
+// It return nil Transaction and nil error when there are no ongoing
+// transactions.
+func (img *SLImages) Transaction(id int) (*Transaction, error) {
+	path := fmt.Sprintf("%s/%d/getTransaction.json", img.block.GetName(), id)
+	p, err := img.client.DoRawHttpRequest(path, "GET", empty)
+	if err != nil {
+		return nil, err
+	}
+
+	if err = newError(p); err != nil {
+		return nil, err
+	}
+
+	if bytes.Compare(p, []byte("null")) == 0 {
+		return nil, nil
+	}
+
+	var t Transaction
+	if err = json.Unmarshal(p, &t); err != nil {
+		return nil, err
+	}
+
+	return &t, nil
+}
+
+// WaitReady waits at most d until all ongoing transactions for image
+// given by the id are finished.
+func (img *SLImages) WaitReady(id int, d time.Duration) error {
+	timeout := time.After(d)
+	parent, err := img.Parent(id)
+	if err != nil {
+		return err
+	}
+
+	if parent != nil && parent.ID != 0 {
+		id = parent.ID
+	}
+
+	children, err := img.Children(id)
+	if err != nil {
+		return err
+	}
+
+	ids := make([]int, len(children))
+	for i, child := range children {
+		ids[i] = child.ID
+	}
+	ids = append(ids, id)
+
+	for {
+		select {
+		case <-timeout:
+			return fmt.Errorf("waiting for %d to be ready timed out after %s", id, d)
+		default:
+			var ongoing bool
+
+			for _, id := range ids {
+				t, err := img.Transaction(id)
+				if err != nil {
+					return err
+				}
+
+				if t != nil && t.ID != 0 {
+					ongoing = true
+					break
+				}
+			}
+
+			if !ongoing {
+				return nil
+			}
+
+			time.Sleep(5 * time.Second)
```

Comment:
I'm don't know how to user ticker to not have to wait the first tick - if there's no transactions in progress the above loop returns right away, with ticker it'll sleep initial 5s regardless of the state.